storagemarket/impl/clientutils/clientutils.go

```diff
@@ -21,21 +22,50 @@ import (
 )
 
 // CommP calculates the commP for a given dataref
-func CommP(ctx context.Context, pieceIO pieceio.PieceIO, rt abi.RegisteredSealProof, data *storagemarket.DataRef, storeID *multistore.StoreID) (cid.Cid, abi.UnpaddedPieceSize, error) {
+// In Markets, CommP = PieceCid.
+// We can't rely on the CARv1 payload in the given CARv2 file being deterministic as the client could have
+// written a "non-deterministic/unordered" CARv2 file.
+// So, we need to do a CARv1 traversal here by giving the traverser a random access CARv2 blockstore that wraps the given CARv2 file.
+func CommP(ctx context.Context, CARv2FilePath string, data *storagemarket.DataRef) (cid.Cid, abi.UnpaddedPieceSize, error) {
+	// if we already have the PieceCid, there's no need to do anything here.
 	if data.PieceCid != nil {
 		return *data.PieceCid, data.PieceSize, nil
 	}
 
+	// It's an error if we don't already have the PieceCid for an offline deal i.e. manual transfer.
 	if data.TransferType == storagemarket.TTManual {
 		return cid.Undef, 0, xerrors.New("Piece CID and size must be set for manual transfer")
 	}
 
-	commp, paddedSize, err := pieceIO.GeneratePieceCommitment(rt, data.Root, shared.AllSelector(), storeID)
+	if CARv2FilePath == "" {
+		return cid.Undef, 0, xerrors.New("need Carv2 file path to get a read-only blockstore")
+	}
+
+	rdOnly, err := blockstore.OpenReadOnly(CARv2FilePath, true)
+	if err != nil {
+		return cid.Undef, 0, xerrors.Errorf("failed to open read-only blockstore :%w", err)
+	}
+	defer rdOnly.Close()
+
+	// do a CARv1 traversal with the DFS selector.
+	sc := car.NewSelectiveCar(ctx, rdOnly, []car.Dag{{Root: data.Root, Selector: shared.AllSelector()}})
+	prepared, err := sc.Prepare()
+	if err != nil {
+		return cid.Undef, 0, xerrors.Errorf("failed to prepare CAR, err=%w", err)
+	}
+
+	// write out the deterministic CARv1 payload to the CommP writer and calculate the CommP.
+	commpWriter := &writer.Writer{}
+	err = prepared.Dump(commpWriter)
+	if err != nil {
+		return cid.Undef, 0, xerrors.Errorf("failed to write CARv1 to commP writer, err=%w", err)
+	}
+	dataCIDSize, err := commpWriter.Sum()
 	if err != nil {
-		return cid.Undef, 0, xerrors.Errorf("generating CommP: %w", err)
+		return cid.Undef, 0, xerrors.Errorf("commpWriter.Sum failed, err=%w", err)
 	}
 
-	return commp, paddedSize, nil
+	return dataCIDSize.PieceCID, dataCIDSize.PieceSize.Unpadded(), nil
```

Comment:
Looks like before we were returning the padded size, and here we're returning unpadded. Just want to confirm this is intentional?

Reply:
@dirkmc 

The variable naming was wrong. Even earlier, we were returning the Unpadded size. Code [here](https://github.com/filecoin-project/go-commp-utils/blob/70bf140d31c73efb08e166994720cbf4766b2a56/pieceio/pieceio.go#L97).